# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/render_target.rs
# Commit: e38dce4a5b06
# Full Hash: e38dce4a5b06869dd4a4d753f41f7594a8cddfb6
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2021-04-22 21:31:57
# Regressor Bug: 1702228
# File Overlap Count: 2
# Description:
#   Bug 1702228 - Cache linear gradients by default. r=gw
#   
#   This patch breaks linear gradients into two primitives:
#    - LinearGradient is always rendered via the brush shader. It is only used with SWGL.
#    - CachedLinearGradient is always rendered via a cached render task and the brush image shader. Its implementation is very similar to conic and radial gradients, with the addition of a fast-path for axis aligned gradients with two stops.
# ==============================================================================

diff -r b3f8befa510f -r e38dce4a5b06 gfx/wr/webrender/src/render_target.rs
--- a/gfx/wr/webrender/src/render_target.rs	Thu Apr 22 10:34:54 2021 +0000
+++ b/gfx/wr/webrender/src/render_target.rs	Thu Apr 22 10:34:54 2021 +0000
@@ -17,7 +17,10 @@
 use crate::internal_types::{FastHashMap, TextureSource, CacheTextureId};
 use crate::picture::{SliceId, SurfaceInfo, ResolvedSurfaceTexture, TileCacheInstance};
 use crate::prim_store::{PrimitiveStore, DeferredResolve, PrimitiveScratchBuffer};
-use crate::prim_store::gradient::{FastLinearGradientInstance, RadialGradientInstance, ConicGradientInstance};
+use crate::prim_store::gradient::{
+    FastLinearGradientInstance, LinearGradientInstance, RadialGradientInstance,
+    ConicGradientInstance,
+};
 use crate::render_backend::DataStores;
 use crate::render_task::{RenderTaskKind, RenderTaskAddress};
 use crate::render_task::{RenderTask, ScalingTask, SvgFilterInfo};
@@ -403,6 +406,7 @@
             RenderTaskKind::Border(..) |
             RenderTaskKind::CacheMask(..) |
             RenderTaskKind::FastLinearGradient(..) |
+            RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
             RenderTaskKind::LineDecoration(..) => {
@@ -498,6 +502,7 @@
             RenderTaskKind::Border(..) |
             RenderTaskKind::LineDecoration(..) |
             RenderTaskKind::FastLinearGradient(..) |
+            RenderTaskKind::LinearGradient(..) |
             RenderTaskKind::RadialGradient(..) |
             RenderTaskKind::ConicGradient(..) |
             RenderTaskKind::SvgFilter(..) => {
@@ -601,6 +606,7 @@
     pub clears: Vec<DeviceIntRect>,
     pub line_decorations: Vec<LineDecorationJob>,
     pub fast_linear_gradients: Vec<FastLinearGradientInstance>,
+    pub linear_gradients: Vec<LinearGradientInstance>,
     pub radial_gradients: Vec<RadialGradientInstance>,
     pub conic_gradients: Vec<ConicGradientInstance>,
 }
@@ -616,6 +622,7 @@
             clears: vec![],
             line_decorations: vec![],
             fast_linear_gradients: vec![],
+            linear_gradients: vec![],
             radial_gradients: vec![],
             conic_gradients: vec![],
         }
@@ -688,6 +695,9 @@
             RenderTaskKind::FastLinearGradient(ref task_info) => {
                 self.fast_linear_gradients.push(task_info.to_instance(&target_rect));
             }
+            RenderTaskKind::LinearGradient(ref task_info) => {
+                self.linear_gradients.push(task_info.to_instance(&target_rect, gpu_cache));
+            }
             RenderTaskKind::RadialGradient(ref task_info) => {
                 self.radial_gradients.push(task_info.to_instance(&target_rect, gpu_cache));
             }